cmd/publicize/main.go

```diff
@@ -0,0 +1,188 @@
+	fs.StringVar(&o.gitEmail, "git-email", "", "The email to use on the git commit. Requires --git-name. If not specified, uses the system default.")
+
+	for _, group := range []flagutil.OptionGroup{&o.github} {
+		group.AddFlags(fs)
+	}
+	o.githubEventServerOptions.Bind(fs)
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatalf("cannot parse args: '%s'", os.Args[1:])
+	}
+	return o
+}
+
+func (o *options) Validate() error {
+	if (o.gitEmail == "") != (o.gitName == "") {
+		return errors.New("--git-name and --git-email must be specified")
+	}
+
+	for _, group := range []flagutil.OptionGroup{&o.github} {
+		if err := group.Validate(o.dryRun); err != nil {
+			return err
+		}
+	}
+
+	bytes, err := ioutil.ReadFile(o.configPath)
+	if err != nil {
+		return fmt.Errorf("Couldn't read publicize configuration file: %v", o.configPath)
+	}
+
+	var c Config
+	if err := yaml.Unmarshal(bytes, &c); err != nil {
+		return fmt.Errorf("Couldn't unmarshal publicize configuration: %w", err)
+	}
+
+	if err := c.validate(); err != nil {
+		return err
+	}
+
+	o.bonds = make(map[string]string)
+	for _, bond := range c.Bonds {
+		o.bonds[bond.From] = bond.To
+	}
+
+	if err := o.githubEventServerOptions.DefaultAndValidate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func main() {
+	o := gatherOptions()
+	if err := o.Validate(); err != nil {
+		logrus.Fatalf("Invalid options: %v", err)
+	}
+
+	secretAgent := &secret.Agent{}
+	if err := secretAgent.Start([]string{o.github.TokenPath, o.webhookSecretFile}); err != nil {
+		logrus.WithError(err).Fatal("Error starting secrets agent.")
+	}
+
+	webhookTokenGenerator := secretAgent.GetTokenGenerator(o.webhookSecretFile)
+	githubTokenGenerator := secretAgent.GetTokenGenerator(o.github.TokenPath)
+	getter := func() sets.String {
+		return sets.NewString([]string{string(webhookTokenGenerator()), string(githubTokenGenerator())}...)
+	}
+	logrus.SetFormatter(logrusutil.NewCensoringFormatter(logrus.StandardLogger().Formatter, getter))
+
+	githubClient, err := o.github.GitHubClient(secretAgent, o.dryRun)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error getting GitHub client.")
+	}
+
+	censor := func(content []byte) []byte {
+		return githubTokenGenerator()
+	}
+	gitClient, err := o.git.GitClient(githubClient, githubTokenGenerator, censor, o.dryRun)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error getting Git client.")
+	}
+
+	logger := logrus.WithField("component", "publicize")
+	serv := &server{
+		githubTokenGenerator: githubTokenGenerator,
+		bonds:                o.bonds,
+		ghc:                  githubClient,
+		gc:                   gitClient,
+		gitName:              o.gitName,
+		gitEmail:             o.gitEmail,
+		remoteHost:           fmt.Sprintf("https://%s", github.DefaultHost),
+		dry:                  o.dryRun,
+	}
+
+	o.githubEventServerOptions.HmacTokenGenerator = webhookTokenGenerator
```

Comment:
These three things seem like something you should get for free, no? What user of the github event server library wouldn't need them?

Reply:
well, we need to pass the options t ocreate a github event server and the hmac, metric and logger are empty in the current state, and we rely on the user to pass them.

Reply:
Right, but if every single user of the github event server will always want these, the constructor should do it for them.

Reply:
Yea I had implemented it this way but it got lost in the reviews. I will open a follow-up PR.

Reply:
https://github.com/kubernetes/test-infra/pull/18426